File: app/src/main/java/br/com/jonathanzanella/myexpenses/database/Where.java
package br.com.jonathanzanella.myexpenses.database;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by jzanella on 11/10/16.
 */

public class Where {
	private class Query {
		Fields field;
		String operation;

		Query(Fields field) {
			this.field = field;
		}

		@Override
		public String toString() {
			return field.toString() + " " + operation + " ?";
		}
	}
	private List<Query> queries;
	private List<String> values;

	public Where(Fields field) {
		queries = new ArrayList<>();
		values = new ArrayList<>();
		queries.add(new Query(field));
	}

	private void setLastQueryOperation(String operation) {
		Query query = queries.get(queries.size() - 1);
		query.operation = operation;
	}

	public Where eq(String s) {
		isExpectingFieldDefinition();
		setLastQueryOperation("=");
		values.add(s);
		return this;
	}

	public Where eq(Long l) {
		isExpectingFieldDefinition();
		setLastQueryOperation("=");
		values.add(String.valueOf(l));
		return this;
	}

	public Where eq(Boolean b) {
		isExpectingFieldDefinition();
		setLastQueryOperation("=");
		values.add(String.valueOf(b ? 1 : 0));
		return this;
	}

	public Where lessThanOrEq(Long l) {
		isExpectingFieldDefinition();
		setLastQueryOperation("<=");
		values.add(String.valueOf(l));
		return this;
	}

	public Where greaterThanOrEq(Long l) {
		isExpectingFieldDefinition();
		setLastQueryOperation(">=");
		values.add(String.valueOf(l));
		return this;
	}

	public Where and(Fields field) {
		fieldsAndValueMatch();
		queries.add(new Query(field));
		return this;
	}

	public Select query() {
		fieldsAndValueMatch();
		StringBuilder query = new StringBuilder();
		for (int i = 0; i < queries.size(); i++) {
			query = query.append(queries.get(i).toString());
			if(i != queries.size() - 1) {
				query = query.append(" and ");
			}
		}
		return new Select(query.toString(), values.toArray(new String[]{}));
	}

	private void fieldsAndValueMatch() {
		if(queries.size() != values.size())
			throw new UnsupportedOperationException("The value for the last field was not setted");
	}

	private void isExpectingFieldDefinition() {
		if(queries.size() - 1 != values.size())
			throw new UnsupportedOperationException("More fields than values added");
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/database/Where.kt
package br.com.jonathanzanella.myexpenses.database

import java.util.*

class Where(field: Fields) {
    private inner class Query internal constructor(internal var field: Fields) {
        internal var operation: String? = null

        override fun toString(): String {
            return field.toString() + " " + operation + " ?"
        }
    }

    private val queries: MutableList<Query>
    private val values: MutableList<String> = ArrayList()
    private var orderBy: Fields = Fields.NAME

    init {
        queries = ArrayList<Query>()
        queries.add(Query(field))
    }

    private fun setLastQueryOperation(operation: String) {
        val query = queries[queries.size - 1]
        query.operation = operation
    }

    fun notEq(s: String): Where {
        isExpectingFieldDefinition()
        setLastQueryOperation("!=")
        values.add(s)
        return this
    }

    fun eq(s: String): Where {
        isExpectingFieldDefinition()
        setLastQueryOperation("=")
        values.add(s)
        return this
    }

    fun eq(l: Long?): Where {
        isExpectingFieldDefinition()
        setLastQueryOperation("=")
        values.add(l.toString())
        return this
    }

    fun eq(b: Boolean?): Where {
        isExpectingFieldDefinition()
        setLastQueryOperation("=")
        values.add((if (b != null && b) 1 else 0).toString())
        return this
    }

    fun lessThanOrEq(l: Long?): Where {
        isExpectingFieldDefinition()
        setLastQueryOperation("<=")
        values.add(l.toString())
        return this
    }

    fun greaterThanOrEq(l: Long?): Where {
        isExpectingFieldDefinition()
        setLastQueryOperation(">=")
        values.add(l.toString())
        return this
    }

    fun queryIn(list: List<String>): Where {
        var string = "("
        for (s in list) {
            string = string.plus("?,")
            values.add(s)
        }
        string = string.removeRange(string.length - 1, string.length) + ")"
        setLastQueryOperation(string)
        return this
    }

    fun and(field: Fields): Where {
        queries.add(Query(field))
        return this
    }

    fun query(): Select {
        var query = StringBuilder()
        for (i in queries.indices) {
            query = query.append(queries[i].toString())
            if (i != queries.size - 1) {
                query = query.append(" and ")
            }
        }
        return Select(query.toString(), values.toTypedArray())
    }

    fun orderBy(field: Fields): Where {
        orderBy = field
        return this
    }

    fun orderBy(): Fields {
        return orderBy
    }

    private fun isExpectingFieldDefinition() {
        if (queries.size - 1 != values.size)
            throw UnsupportedOperationException("More fields than values added")
    }
}